### PR TITLE
Allow Crashes to track error with user-defined custom properties

### PR DESCRIPTION
**Pull request by dhei**

### CODE REVIEW

File: AppCenterCrashes/AppCenterCrashes/MSCrashes.mm [hunk 1]
```diff
@@ -44,8 +44,34 @@
 
 static unsigned int kMaxAttachmentsPerCrashReport = 2;
 
+/**
+ * Maxinum properties per event
```
Comment (bmourat) [inline AppCenterCrashes/AppCenterCrashes/MSCrashes.mm]: Small typo here: Maximum

File: AppCenterCrashes/AppCenterCrashes/MSCrashes.mm [hunk 2]
```diff
@@ -44,8 +44,34 @@
...
+ */
+static const int maxPropertiesPerEvent = 5;
+
+/**
+ * Mininum properties key length
```
Comment (bmourat) [inline AppCenterCrashes/AppCenterCrashes/MSCrashes.mm]: Same here and couple below

File: AppCenterCrashes/AppCenterCrashes/MSCrashes.mm [hunk 3]
```diff
@@ -1210,18 +1244,75 @@ - (void)notifyWithUserConfirmation:(MSUserConfirmation)userConfirmation {
...
+
 #pragma mark - Handled exceptions
 
-- (void)trackModelException:(MSException *)exception {
+- (void)trackModelException:(MSException *)exception
+             withProperties:(NSDictionary<NSString *, NSString *> *)properties {
   if (![self isEnabled])
-    return;
-
+      return;
```
Comment (bmourat) [inline AppCenterCrashes/AppCenterCrashes/MSCrashes.mm]: Wrong indentation

File: AppCenterCrashes/AppCenterCrashes/MSCrashes.mm [hunk 4]
```diff
@@ -1210,18 +1244,75 @@ - (void)notifyWithUserConfirmation:(MSUserConfirmation)userConfirmation {
...
+  
   // Create an error log.
   MSHandledErrorLog *log = [MSHandledErrorLog new];
-
+  
   // Set properties of the error log.
   log.errorId = MS_UUID_STRING;
   log.exception = exception;
+  if (properties && properties.count > 0) {
+    // Send only valid properties.
```
Comment (MatkovIvan) [inline AppCenterCrashes/AppCenterCrashes/MSCrashes.mm]: Empty line before comment

File: AppCenterCrashes/AppCenterCrashesTests/MSCrashesTests.mm [hunk 5]
```diff
@@ -651,19 +654,161 @@ - (void)testTrackModelException {
...
+     @"Key5" : @"" };
+  
+  // When
+  NSDictionary *validatedProperties =
+  [[MSCrashes sharedInstance] validateProperties:validProperties andType:kMSTypeHandledError];
+  
+  // Then
+  XCTAssertTrue([validatedProperties count] == [validProperties count]);
+  
+  // Test too many properties in one event
```
Comment (MatkovIvan) [inline AppCenterCrashes/AppCenterCrashesTests/MSCrashesTests.mm]: Missing dot. Same in the comments below.

**Comment by dhei** [comment]
FYI to @yimysty

File: AppCenterCrashes/AppCenterCrashes/MSCrashes.mm [hunk 6]
```diff
@@ -44,8 +44,34 @@
 
 static unsigned int kMaxAttachmentsPerCrashReport = 2;
 
+/**
+ * Maximum properties per event
+ */
+static const int maxPropertiesPerEvent = 5;
```
Comment (jaelim-ms) [inline AppCenterCrashes/AppCenterCrashes/MSCrashes.mm]: Event? Please rename.

File: AppCenterCrashes/AppCenterCrashes/MSCrashes.mm [hunk 7]
```diff
@@ -1210,18 +1244,76 @@ - (void)notifyWithUserConfirmation:(MSUserConfirmation)userConfirmation {
...
+
 #pragma mark - Handled exceptions
 
-- (void)trackModelException:(MSException *)exception {
+- (void)trackModelException:(MSException *)exception
+             withProperties:(NSDictionary<NSString *, NSString *> *)properties {
   if (![self isEnabled])
     return;
-
+
```
Comment (jaelim-ms) [inline AppCenterCrashes/AppCenterCrashes/MSCrashes.mm]: Please run code formatter.

File: AppCenterCrashes/AppCenterCrashesTests/MSCrashesTests.mm [hunk 8]
```diff
@@ -651,19 +654,161 @@ - (void)testTrackModelException {
         exception = log.exception;
       });
   [MSAppCenter configureWithAppSecret:kMSTestAppSecret];
-  [self.sut startWithLogManager:logManagerMock appSecret:kMSTestAppSecret];
```
Comment (jaelim-ms) [inline AppCenterCrashes/AppCenterCrashesTests/MSCrashesTests.mm]: Any reason of this change?

File: AppCenterCrashes/AppCenterCrashes/MSCrashes.mm (hunk 6 above)
Comment (dhei) [reply AppCenterCrashes/AppCenterCrashes/MSCrashes.mm]: @jaelim-ms good catch, thanks.

File: AppCenterCrashes/AppCenterCrashesTests/MSCrashesTests.mm (hunk 8 above)
Comment (dhei) [reply AppCenterCrashes/AppCenterCrashesTests/MSCrashesTests.mm]: @jaelim-ms In order to be consistent with other non-public APIs (like [`trackPage` of Analytics](https://github.com/Microsoft/AppCenter-SDK-Apple/blob/develop/AppCenterAnalytics/AppCenterAnalytics/Internals/MSAnalyticsInternal.h#L24)), I changed method signature of `- (void)trackModelException:(MSException *)exception;` to static `+ (void)trackModelException:(MSException *)exception;` in `MSCrashesInternal.h`. 

As a consequence, in unit test we have to change this line to `[[MSCrashes sharedInstance] startWithLogManager:logManagerMock appSecret:kMSTestAppSecret];`.

**Pull request merged**